### PR TITLE
leveldb: make heavy version reference asynchronous

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR introduces an efficient approach for table file reference.

The original for file reference is:
* When the database version is first time created, then references all relative table files by increasing file ref by 1
* When the version is released(no operation hold the version), deferences all  table files by decreasing file ref by 1

If the database size is 1TB with each file size is 2MB, image the cost for 500,000 file reference cost.
And worse, the read operation(also iterator) will also acquire a version before reading. And it has **very high possibility** then this read operation is the last thing holding the version, it means the cost of huge file deferences is distributed to the read operation. **It will definitely slow down the entire system**.

Therefore I introduce an efficient approach for it. Notably, the new version is ONLY created by compaction. While luckily we have the `changed file set` for every compaction(no matter mem_compaction or table_compaction). All we need to is "delay reference opt and wait for release opt, use the `changed file set` to delete useless files and reference new generated files".

Two things we need to care about:
* We have to process version reference and deference one by one, otherwise we can delete some files that are referencing now.
* Some operation like iterator can hold a version for a very long time. It means such operation can block the entire processing queue. In order to break it, we need to convert some version task to normal mode(references all files and also deferences)

### CODE REVIEW

File: leveldb/session_util.go [hunk 1]
```diff
@@ -39,20 +39,155 @@ func (s *session) newTemp() storage.FileDesc {
...
+			referenced[next] = struct{}{}
+			delete(ref, next)
+			delete(deltas, next)
+			next += 1
+		}
+
+		// Use delta information to process all released versions.
+		for {
+			if d, exist := released[next]; exist {
+				if d != nil {
```
Comment (holiman) [inline leveldb/session_util.go]: can `d` be `nil` if `exist` is true?

Comment (rjl493456442) [reply leveldb/session_util.go]: Yes, for `read` operation, it just references the corresponding version and then releases when read is done. No `delta` introduced.

**Comment by rjl493456442** [comment]
@syndtr Any chance to take a look :)

**Comment by karalabe** [comment]
Some benchmark results from running this (and Gary's other PR) on top of our insane workload (the horizontal line is when we deployed this PR):

Screenshot from 2019-03-13 10-31-37

First up, even though our VMs have 500MB/s read/write SSDs, our system failed to (consistently) push them over 50MB/s prior to this PR. This PR manages to load them up to around 350MB/s, and also keep that load. If we were to ignore every other load our system does, this alone would mean that this PR has about a 7x (!!) performance gain.

Screenshot from 2019-03-13 10-31-43

Looking at out own domain specific usage of go-leveldb, we can see that our processing performance, which was fluctuating like crazy between 750ms and 2s on average per task is now flattened to a 250-350ms range, essentially confirming a 3-5x performance increase.

**Comment by syndtr** [comment]
Well done @rjl493456442, I'll definitely looks into it. @karalabe nice.

**Comment by syndtr** [comment]
> Some operation like iterator can hold a version for a very long time. It means such operation can block the entire processing queue. In order to break it, we need to convert some version task to normal mode(references all files and also deferences)

@rjl493456442 So this is triggered by the number of backlog right? How about adding timeout trigger to it as well?
Database that not that write-heavy or only do writes on occasion may take long time to reach the threshold.

**Comment by rjl493456442** [comment]
@syndtr Yes, it totally makes sense. In the beginning, this mechanism is aimed to fix the memory issue if there are too many pending tasks. Timeout trigger can ensure each version task has a fixed time cycle.

**Comment by rjl493456442** [comment]
@syndtr Updated. Please take another look.

**Comment by syndtr** [comment]
@rjl493456442 Also add timer, to wake-up the task loop?

**Comment by rjl493456442** [comment]
@syndtr added.

**Comment by syndtr** [comment]
Merged! Thanks @rjl493456442

**Pull request merged**